File: src/lib/api/knowledge-domain/qa.ts
import { save as saveFileDialog } from "@tauri-apps/plugin-dialog";
import { knowledgeApi } from "../knowledge";
import type { KnowledgeAskResult } from "@/types";
import type {
  KnowledgeQaSession,
  KnowledgeQaSessionDetail,
  KnowledgeScopedQuestionInput,
  PersistKnowledgeQaRoundInput,
} from "@/types/knowledge-domain/qa";
import { devApiFetch, hasTauriRuntime, invoke } from "../client";

export type KnowledgeQaSaveResult = "saved" | "downloaded" | "cancelled";

export const knowledgeQaApi = {
  previewContext: knowledgeApi.previewRagContext,
  ask: knowledgeApi.ask,
  runEvaluation: knowledgeApi.runFixedRetrievalEvaluation,
  askScopedQuestion: (input: KnowledgeScopedQuestionInput) =>
    hasTauriRuntime()
      ? invoke<KnowledgeAskResult>("ask_knowledge_scoped_question", { input })
      : devApiFetch<KnowledgeAskResult>(
          `/knowledge/projects/${input.projectId}/qa/ask`,
          { method: "POST", body: JSON.stringify(input) },
        ),
  listSessions: (projectId: number) =>
    hasTauriRuntime()
      ? invoke<KnowledgeQaSession[]>("list_knowledge_qa_sessions", {
          projectId,
        })
      : devApiFetch<KnowledgeQaSession[]>(
          `/knowledge/projects/${projectId}/qa/sessions`,
        ),
  getSession: (projectId: number, sessionId: number) =>
    hasTauriRuntime()
      ? invoke<KnowledgeQaSessionDetail>("get_knowledge_qa_session", {
          projectId,
          sessionId,
        })
      : devApiFetch<KnowledgeQaSessionDetail>(
          `/knowledge/projects/${projectId}/qa/sessions/${sessionId}`,
        ),
  persistRound: (input: PersistKnowledgeQaRoundInput) =>
    hasTauriRuntime()
      ? invoke<KnowledgeQaSessionDetail>("persist_knowledge_qa_round", {
          input,
        })
      : devApiFetch<KnowledgeQaSessionDetail>(
          `/knowledge/projects/${input.projectId}/qa/rounds`,
          { method: "POST", body: JSON.stringify(input) },
        ),
  deleteSession: (projectId: number, sessionId: number) =>
    hasTauriRuntime()
      ? invoke<void>("delete_knowledge_qa_session", {
          projectId,
          sessionId,
        })
      : devApiFetch<void>(
          `/knowledge/projects/${projectId}/qa/sessions/${sessionId}`,
          { method: "DELETE" },
        ),
  saveMarkdown: async (input: {
    content: string;
    defaultFileName: string;
  }): Promise<KnowledgeQaSaveResult> => {
    const safeFileName = input.defaultFileName.replace(/[\\/:*?"<>|]/g, "_");
    if (hasTauriRuntime()) {
      const path = await saveFileDialog({
        defaultPath: safeFileName.endsWith(".md")
          ? safeFileName
          : `${safeFileName}.md`,
        filters: [{ name: "Markdown", extensions: ["md", "markdown"] }],
      });
      if (!path) return "cancelled";
      await invoke<string>("save_knowledge_qa_markdown", {
        input: { path, content: input.content },
      });
      return "saved";
    }

    // 浏览器 Dev API 没有本地文件插件时仍提供手动下载，便于验收和导出草稿。
    const blob = new Blob([input.content], {
      type: "text/markdown;charset=utf-8",
    });
    const url = URL.createObjectURL(blob);
    const link = document.createElement("a");
    link.href = url;
    link.download = safeFileName.endsWith(".md")
      ? safeFileName
      : `${safeFileName}.md`;
    link.rel = "noopener";
    link.style.display = "none";
    document.body.appendChild(link);
    link.click();
    link.remove();
    window.setTimeout(() => URL.revokeObjectURL(url), 0);
    return "downloaded";
  },
};

File: src-tauri/src/commands/knowledge_domain/qa.rs
use tauri::Manager;

use crate::error::{AppError, CommandError};
use serde::Deserialize;

use crate::models::knowledge_domain::qa::{
    KnowledgeQaSession, KnowledgeQaSessionDetail, KnowledgeScopedQuestionInput,
    PersistKnowledgeQaRoundInput,
};
use crate::models::KnowledgeAskResult;
use crate::services::knowledge_domain::qa::KnowledgeScopedQuestionService;
use crate::services::knowledge_domain::qa_export::KnowledgeQaExportService;
use crate::state::AppState;

pub(crate) const DOMAIN: &str = "qa";

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveKnowledgeQaMarkdownInput {
    pub path: String,
    pub content: String,
}

/// 项目问答必须在后端从路由范围构造检索条件；页面无法传入全局项目或任意本地路径。
#[tauri::command]
pub async fn ask_knowledge_scoped_question(
    app: tauri::AppHandle,
    state: tauri::State<'_, AppState>,
    input: KnowledgeScopedQuestionInput,
) -> Result<KnowledgeAskResult, CommandError> {
    let app_data_dir = app
        .path()
        .app_data_dir()
        .map_err(|error| AppError::Custom(error.to_string()))?;
    KnowledgeScopedQuestionService::ask(&state.db, &app_data_dir, input)
        .await
        .map_err(Into::into)
}

#[tauri::command]
pub fn list_knowledge_qa_sessions(
    state: tauri::State<'_, AppState>,
    project_id: i64,
) -> Result<Vec<KnowledgeQaSession>, CommandError> {
    KnowledgeScopedQuestionService::list_sessions(&state.db, project_id).map_err(Into::into)
}

#[tauri::command]
pub fn get_knowledge_qa_session(
    state: tauri::State<'_, AppState>,
    project_id: i64,
    session_id: i64,
) -> Result<KnowledgeQaSessionDetail, CommandError> {
    KnowledgeScopedQuestionService::get_session(&state.db, project_id, session_id)
        .map_err(Into::into)
}

#[tauri::command]
pub fn persist_knowledge_qa_round(
    state: tauri::State<'_, AppState>,
    input: PersistKnowledgeQaRoundInput,
) -> Result<KnowledgeQaSessionDetail, CommandError> {
    KnowledgeScopedQuestionService::persist_round(&state.db, input).map_err(Into::into)
}

#[tauri::command]
pub fn delete_knowledge_qa_session(
    state: tauri::State<'_, AppState>,
    project_id: i64,
    session_id: i64,
) -> Result<(), CommandError> {
    KnowledgeScopedQuestionService::delete_session(&state.db, project_id, session_id)
        .map_err(Into::into)
}

/// 保存用户手动触发的 Markdown 问答记录；文件内容由 Service 统一脱敏并原子写入。
#[tauri::command]
pub fn save_knowledge_qa_markdown(
    input: SaveKnowledgeQaMarkdownInput,
) -> Result<String, CommandError> {
    KnowledgeQaExportService::save_markdown(&input.path, &input.content).map_err(Into::into)
}
